File: code/735.asteroid-collision.php
<?php
/*
 * @lc app=leetcode id=735 lang=php
 *
 * [735] Asteroid Collision
 */

// @lc code=start
class Solution {

    /**
     * @param Integer[] $asteroids
     * @return Integer[]
     */
    function asteroidCollision($asteroids) {
        $result = [];
        foreach ($asteroids as $asteroid) {
            if ($asteroid > 0) {
                $result[] = $asteroid;
            } else {
                $this->collide($result, $asteroid);
            }
        }

        return $result;
    }

    function collide(&$leftAsteroids, $current) {
        $lastKey = array_key_last($leftAsteroids);
        if (!isset($lastKey)) {
            $leftAsteroids[] = $current;
            return;
        }

        $left = $leftAsteroids[$lastKey];
        if ($left < 0) {
            $leftAsteroids[] = $current;
            return;
        }

        if (abs($left) == abs($current)) {
            unset($leftAsteroids[$lastKey]);
        } elseif (abs($left) < abs($current)) {
            unset($leftAsteroids[$lastKey]);
            $this->collide($leftAsteroids, $current);
        } else {
            return;
        }
    }
}
// @lc code=end

File: code/735.asteroid-collision.go
package code

/*
 * @lc app=leetcode id=735 lang=golang
 *
 * [735] Asteroid Collision
 */

// @lc code=start
func asteroidCollision(asteroids []int) []int {
	stack := make([]int, 0)
	stack = append(stack, asteroids[0])

	for i := 1; i < len(asteroids); i++ {
		if len(stack) == 0 {
			stack = append(stack, asteroids[i])
			continue
		}

		if asteroids[i] > 0 {
			stack = append(stack, asteroids[i])
			continue
		}

		if stack[len(stack)-1] < 0 {
			stack = append(stack, asteroids[i])
			continue
		}

		for {
			if len(stack) == 0 || stack[len(stack)-1] < 0 {
				stack = append(stack, asteroids[i])
				break
			}

			if a := stack[len(stack)-1]; a+asteroids[i] > 0 {
				break
			} else if a+asteroids[i] == 0 {
				stack = stack[:len(stack)-1]
				break
			} else {
				stack = stack[:len(stack)-1]
			}
		}
	}

	return stack
}

// @lc code=end
